lib/src/rrule.dart

```diff
@@ -0,0 +1,79 @@
+import 'package:meta/meta.dart';
+
+import 'cache.dart';
+import 'recurrence_rule.dart';
+import 'utils.dart';
+
+class RRule {
+  RRule({required this.rule, required this.start, Cache? initialCache}) {
+    if (initialCache != null) _cache = initialCache;
+  }
+
+  final RecurrenceRule rule;
+
+  final DateTime start;
+
+  Cache _cache = Cache();
+
+  @visibleForTesting
+  Cache get cache => _cache;
+
+  List<DateTime> before({
+    required DateTime date,
+    bool inc = false,
+  }) {
+    final instances = rule.getInstances(start: start);
+    final argumentsHash = hashList([instances, date, inc]);
```

Comment:
Maybe it can be better?

```suggestion
    final argumentsHash = hashList([rule, start, date, inc]);
```